src/dune_pkg/toolchain.ml

```diff
@@ -0,0 +1,245 @@
+open! Import
+open! Stdune
+
+module Make = struct
+  let path =
+    lazy
+      (* TODO: prefer gmake over make if it's available for compatibility with the BSDs *)
+      (match Bin.which ~path:(Env_path.path Env.initial) "make" with
+       | Some p -> p
+       | None ->
+         User_error.raise
+           ~hints:[ Pp.text "Install \"make\" with your system package manager." ]
+           [ Pp.text
+               "The program \"make\" does not appear to be installed. This program is \
+                needed to compile the ocaml toolchain."
+           ])
+  ;;
+end
+
+module Dir = struct
+  let make_and_check_dir path =
+    if not (Path.exists path) then Path.mkdir_p path;
+    if not (Path.is_directory path)
+    then
+      User_error.raise
+        [ Pp.textf "Expected %s to be a directory but it is not." (Path.to_string path) ]
+  ;;
+
+  let cache_dir () =
+    let env_get = Env.get Env.initial in
+    let d =
+      match env_get "XDG_CACHE_HOME" with
+      | Some dir -> Path.of_string dir
+      | None ->
+        (match env_get "HOME" with
+         | Some dir -> Path.relative (Path.of_string dir) ".cache"
+         | None ->
+           User_error.raise
+             [ Pp.text
+                 "Unable to find directory of dune toolchain as neither $XDG_CACHE_HOME \
+                  nor $HOME environment variables are set."
+             ]
+             ~hints:
+               [ Pp.text
+                   "Make sure at least one of $XDG_CACHE_HOME and $HOME are set in your \
+                    environment."
+               ])
+    in
+    make_and_check_dir d;
+    d
+  ;;
+
+  let toolchain_base_dir () =
+    let d = Path.relative (Path.relative (cache_dir ()) "dune") "toolchains" in
+    make_and_check_dir d;
+    d
+  ;;
+end
+
+module Version = struct
+  type t =
+    | V_4_14_2
+    | V_5_1_1
+
+  let equal a b =
+    match a, b with
+    | V_4_14_2, V_4_14_2 | V_5_1_1, V_5_1_1 -> true
+    | _ -> false
+  ;;
+
+  let all_by_string = [ "4.14.2", V_4_14_2; "5.1.1", V_5_1_1 ]
+  let to_string t = List.find_exn all_by_string ~f:(fun (_, t_) -> equal t t_) |> fst
+
+  let toolchain_dir t =
+    let d = Path.relative (Dir.toolchain_base_dir ()) (to_string t) in
+    Dir.make_and_check_dir d;
+    d
+  ;;
+
+  let source_dir t = Path.relative (toolchain_dir t) "source"
+  let target_dir t = Path.relative (toolchain_dir t) "target"
+  let bin_dir t = Path.relative (target_dir t) "bin"
+end
+
+module Compiler_package = struct
+  type t =
+    { version : Version.t
+    ; url : OpamUrl0.t
+    ; checksum : Checksum.t
+    }
+
+  let v_4_14_2 =
+    lazy
```

Comment:
It's to prevent `OpamUrl0.of_string` and `Checksum.of_string` from being called when dune starts. This way that work is only done when their value is needed. I thought about making it a thunk instead but using `lazy` means that the value is cached and reused after the first computation.

Reply:
I doubt that this is a lot of computation but a good point not to make people pay the cost of this when most of the time users will not hit this code.